### PR TITLE
Add a mutex for shared Inference Engine plugins

### PR DESCRIPTION
**Pull request by dkurt**
resolves https://github.com/opencv/opencv/issues/13715
resolves https://github.com/opencv/opencv/issues/13569

```
force_builders=Custom

docker_image:Custom=ubuntu-openvino-2018r5:16.04
test_opencl:Custom=ON
buildworker:Custom=linux-2
#test_modules=dnn
```

### CODE REVIEW

File: modules/dnn/src/op_inf_engine.cpp [hunk 1]
```diff
@@ -622,6 +622,7 @@ void InfEngineBackendNet::init(int targetId)
 
 #endif  // IE < R5
 
+static Mutex sharedPluginsMutex;
 static std::map<InferenceEngine::TargetDevice, InferenceEngine::InferenceEnginePluginPtr> sharedPlugins;
```
Comment (alalek) [inline modules/dnn/src/op_inf_engine.cpp]: It is better to wrap global "static" variables into static functions (to ensure initialization of them at least).

Example:
https://github.com/opencv/opencv/blob/3.4.5/modules/dnn/src/dnn.cpp#L94-L98

Comment (dkurt) [reply modules/dnn/src/op_inf_engine.cpp]: @alalek, Is that Ok to use a mutex from `getInitializationMutex()`?

https://github.com/opencv/opencv/blob/e82e672a933c8bd713ff53ac9ca553ede169b566/modules/dnn/src/init.cpp#L52-L57

Comment (alalek) [reply modules/dnn/src/op_inf_engine.cpp]: I believe you can reuse it (declared in dnn's precomp.hpp).

File: modules/dnn/test/test_layers.cpp [hunk 2]
```diff
@@ -39,6 +39,10 @@
 //
 //M*/
 
+#ifdef HAVE_INF_ENGINE
+#include <thread>
+#endif
```
Comment (alalek) [inline modules/dnn/test/test_layers.cpp]: Please keep "test_precomp.hpp" as the first "include" statement.

File: modules/dnn/src/op_inf_engine.cpp [hunk 3]
```diff
@@ -622,14 +622,20 @@ void InfEngineBackendNet::init(int targetId)
...
+    return sharedPlugins;
+}
 
 void InfEngineBackendNet::initPlugin(InferenceEngine::ICNNNetwork& net)
 {
     CV_Assert(!isInitialized());
 
     try
     {
+        AutoLock lock(getInitializationMutex());
```
Comment (shahzadlone) [inline modules/dnn/src/op_inf_engine.cpp]: perhaps const if not modified.

Comment (dkurt) [reply modules/dnn/src/op_inf_engine.cpp]: @shahzadlone, thanks for review! Could you please clarify it a bit?

Comment (shahzadlone) [reply modules/dnn/src/op_inf_engine.cpp]: if lock isn't being modified could it perhaps be made AutoLock const?

Comment (alalek) [reply modules/dnn/src/op_inf_engine.cpp]: This change doesn't fix (or change) anything in current code.

@shahzadlone Please report about real issues only.

Avoid reporting about "const" of local variables. They are usually not required and can be safely omitted (compilers are smart enough).
It make sense to report about "const" in parameters or about "const methods".

**Pull request merged**